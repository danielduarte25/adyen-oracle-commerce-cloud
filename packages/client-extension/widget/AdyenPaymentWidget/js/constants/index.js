export const baseAPIUrl = '/ccstorex/custom/adyen/v1'
export const channels = { preview: 'preview', storefront: 'storefront' }
export const errorMessages = {
    paymentRefused: 'paymentRefused',
    installmentsConfiguration: 'installmentsConfiguration',
    failed3dsValidation: 'failed3dsValidation',
}
export const paymentMethodTypes = {
    scheme: 'scheme',
    boleto: 'boletobancario',
    local: 'local',
}
export const adyenUrl = (env) => `https://checkoutshopper-${env.toLowerCase()}.adyen.com`
export const adyenCssUrl = (env) => `${adyenUrl(env)}/checkoutshopper/sdk/3.4.0/adyen.css`

export const locale = 'locale'
export const countries = {
    brazil: { locale: 'pt_br', currency: 'brl', countryCode: 'br' },
    mexico: { locale: 'es_mx', currency: 'mxn', countryCode: 'mx' },
}
export const brazilEnabled = 'brazilEnabled'
export const comboCards = { debit: 'debitCard', credit: 'creditCard' }
export const bins = { electron: 'electron', maestro: 'maestro', elodebit: 'elodebit' }
export const storage = { paymentData: 'AdyenPaymentData', order: 'AdyenOrder' }
export const noInstallmentsMsg = 'noInstallmentsMsg'

export const render = 'render'
export const id = 'id'
export const order = 'order'
export const checkout = {
    card: 'checkoutCard',
    boleto: 'checkoutBoleto',
    local: 'checkoutLocal',
}
export const isLoaded = 'isLoaded'
export const ajax = 'ajax'
export const installments = 'installments'
export const cart = 'cart'
export const user = 'user'
export const setPayment = 'setPayment'
export const initialOrderCreated = 'initialOrderCreated'
export const pageChanged = 'pageChanged'
export const installmentsEnabled = 'installmentsEnabled'
export const isAllowedCountryForInstallments = 'isAllowedCountryForInstallments'
export const environment = 'environment'
export const selectedBrand = 'selectedBrand'
export const selectedComboCard = 'selectedComboCard'
export const installmentsOptions = 'installmentsOptions'
export const storedPaymentType = 'storedPaymentType'
export const storedPaymentMethods = 'storedPaymentMethods'
export const paymentMethodsResponse = 'paymentMethodsResponse'
export const translate = 'translate'
export const isPaymentStored = 'isPaymentStored'
export const paymentDetails = 'paymentDetails'
export const selectedInstallment = 'selectedInstallment'
export const comboCardOptions = 'comboCardOptions'
export const genericPayment = 'genericPayment'
export const isDone = 'isDone'
export const isValid = 'isValid'
export const pageParams = 'pageParams'
export const presentToShopper = 'PresentToShopper'
export const orderPayload = 'orderPayload'
export const boletoOptions = {
    deliveryDate: 'boletoDeliveryDate',
    shopperStatement: 'boletoShopperStatement',
}
export const localPaymentMethods = 'localPaymentMethods'
export const card = 'card'
export const clientKey = 'clientKey'
export const holderNameEnabled = 'holderNameEnabled'
export const checkoutComponent = 'checkoutComponent'
export const countryCode = 'countryCode'
export const additionalDetails = 'additionalDetails'
