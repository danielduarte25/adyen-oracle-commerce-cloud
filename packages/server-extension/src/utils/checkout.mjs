import AdyenApiLibrary from '@adyen/api-library'
const { CheckoutAPI } = AdyenApiLibrary

function getCheckout(req) {
    const { client } = req.app.locals
    const checkout = new CheckoutAPI(client)

    return checkout
}

export function getExternalProperties(properties) {
    const additionalProperties = Object.entries(properties).reduce(
        (acc, [key, value]) => ({
            ...acc,
            [key]: JSON.stringify(value),
        }),
        {}
    )
    const externalProperties = Object.keys(properties)

    return {
        additionalProperties,
        externalProperties,
    }
}

export default getCheckout
