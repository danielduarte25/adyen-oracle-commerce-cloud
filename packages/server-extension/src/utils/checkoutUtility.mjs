import { CheckoutUtility } from '@adyen/api-library'

function getCheckoutUtility(req) {
    const { client } = req.app.locals
    const checkout = new CheckoutUtility(client)

    return checkout
}

export default getCheckoutUtility
