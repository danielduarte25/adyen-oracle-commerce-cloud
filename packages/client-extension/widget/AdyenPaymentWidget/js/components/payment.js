import * as constants from '../constants'
import { eventEmitter } from '../utils'
import { store } from './index'

class Payment {
    constructor() {
        this.startEventListeners()
    }

    startEventListeners = () => {
        eventEmitter.payment.on(constants.setPayment, this.setPayment)
    }

    setPayment = (type) => {
        const order = store.get(constants.order)
        const id = store.get(constants.id)

        const genericPayment = store.get(constants.genericPayment)
        const updatedGenericPayment = { ...genericPayment, id: id() }

        eventEmitter.store.emit(constants.genericPayment, updatedGenericPayment)
        eventEmitter.component.emit(constants.comboCardOptions)

        const payments = [updatedGenericPayment]

        order().updatePayments(payments)

        this.updatePayment(this.createPayment(type))
    }

    updatePayment = (pPayment) => {
        const order = store.get(constants.order)
        const id = store.get(constants.id)

        const payments = order().payments()
        const checkId = (curId, id) => curId && curId === id()
        const setPayments = (payment, index) => {
            const curPayment = payments[index]
            const { id: curId } = curPayment
            const isSameId = checkId(curId, id)
            return isSameId ? pPayment : curPayment
        }

        const newPayments = payments.map(setPayments)
        order().payments(newPayments)
    }

    updatePaymentMethodType = (type, paymentDetails) => {
        const brandType = this.getBrandType(type)
        const paymentMethod = paymentDetails.paymentMethod && { ...paymentDetails.paymentMethod, ...brandType }
        const payload = { ...paymentDetails, paymentMethod }
        return payload
    }

    getBasketItems = () => {
        const cart = store.get(constants.cart)
        const getItem = (acc, item, idx) => ({
            ...acc,
            [`item${idx + 1}`]: {
                itemID: item.commerceItemId,
                quantity: item.quantity(),
                productTitle: item.productData().displayName,
                category: item.productData().parentCategory.repositoryId,
                brand: item.productData().brand,
                sku: item.catRefId,
            },
        })
        return cart().items().reduce(getItem, {})
    }

    getAccountInfo = () => {
        const user = store.get(constants.user)
        return {
            accountCreationDate: user().registrationDate,
        }
    }

    createPayment = (type) => {
        const storedPaymentType = store.get(constants.storedPaymentType)
        const isPaymentStored = store.get(constants.isPaymentStored)
        const paymentDetails = store.get(constants.paymentDetails)
        const selectedInstallment = store.get(constants.selectedInstallment)
        const selectedComboCard = store.get(constants.selectedComboCard)
        const comboCardOptions = store.get(constants.comboCardOptions)
        const additionalDetails = store.get(constants.additionalDetails)

        const hasInstallments = this.checkSelectedInstallment(selectedInstallment)

        const isCreditCard = selectedComboCard() === constants.comboCards.credit
        const isValid = this.validateInstallment(hasInstallments, isCreditCard)

        const storedPayment = this.getStoredPayment(isPaymentStored, storedPaymentType)
        const numberOfInstallments = this.getInstallments(isValid, selectedInstallment)

        const genericPayment = store.get(constants.genericPayment)
        const updatedPaymentDetails = this.updatePaymentMethodType(type, paymentDetails[type])
        const basketItems = this.getBasketItems()
        const accountInfo = this.getAccountInfo()

        const customProperties = {
            accountInfo,
            paymentDetails: updatedPaymentDetails,
            storedPayment,
            riskData: {
                basket: basketItems,
            },
            ...additionalDetails,
            ...(numberOfInstallments && { numberOfInstallments }),
            ...comboCardOptions,
        }
        const updatedGenericPayment = { ...genericPayment, customProperties }

        return updatedGenericPayment
    }

    checkSelectedInstallment = (selectedInstallment) =>
        selectedInstallment() && 'numberOfInstallments' in selectedInstallment()
    getInstallments = (isValid, selectedInstallment) => isValid && parseInt(selectedInstallment().numberOfInstallments)
    validateInstallment = (hasInstallments, isCreditCard) => hasInstallments && isCreditCard
    getStoredPayment = (isPaymentStored, storedPaymentType) => (isPaymentStored ? storedPaymentType() : '')

    getBrandType = (type) => {
        const selectedComboCard = store.get(constants.selectedComboCard)()
        const isDebitCard = selectedComboCard === constants.comboCards.debit
        const brand = store.get(constants.selectedBrand)
        const isGeneric = type === constants.paymentMethodTypes.scheme
        const isComboCard = isGeneric && isDebitCard

        const brandType = isComboCard && { type: brand }
        return brandType
    }
}

export default Payment
