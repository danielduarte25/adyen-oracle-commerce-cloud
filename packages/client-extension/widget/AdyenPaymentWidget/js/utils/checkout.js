import ccConstants from 'ccConstants'
import AdyenCheckout from '@adyen/adyen-web'
import '@adyen/adyen-web/dist/adyen.css'
import { store } from '../components'
import * as constants from '../constants'
import { eventEmitter } from './index'
import { redirectAuth, getOrderPayload } from '../utils'

export const getDefaultConfig = (type) => {
    const cart = store.get(constants.cart)
    const { amount, currencyCode } = cart()
    const environment = store.get(constants.environment)
    const countryCode = store.get(constants.countryCode)
    const clientKey = store.get(constants.clientKey)
    const paymentMethodsResponse = store.get(constants.paymentMethodsResponse)

    return {
        amount: {
            value: amount() * 100,
            currency: currencyCode(),
        },
        countryCode,
        showPayButton: true,
        environment: environment.toLowerCase(),
        clientKey,
        paymentMethodsResponse,
        onAdditionalDetails: (state, component) => {
            console.log('on additional details')
            const payment = { type: 'generic', customProperties: state.data }
            eventEmitter.order.emit('recreateOrder', payment)
        },
    }
}

export const createFromAction = ({ action, selector, checkoutComponent }) => {
    checkoutComponent.createFromAction(action).mount(selector)
}

class Checkout {
    constructor(type) {
        this.type = type
    }

    getCheckout = () => {
        const checkout = store.get(constants.checkoutComponent)
        if (checkout) {
            return checkout
        }

        // eslint-disable-next-line no-undef
        const newCheckout = new AdyenCheckout(getDefaultConfig(this.type))
        eventEmitter.store.emit(constants.checkoutComponent, newCheckout)
        return newCheckout
    }

    createCheckout = ({ configuration, selector, type, options = {} }, cb) => {
        const checkout = this.getCheckout()
        try {
            checkout.create(type, { ...options, ...configuration }).mount(selector)
            cb && cb(checkout)
        } catch (e) {
            console.warn(`Payment method not supported: ${type}`)
        }
    }

    onSubmit = (onChange) => (state, component) => {
        onChange && onChange(state, component)
        const loader = document.querySelector(`.loader-wrapper`)
        loader && loader.classList.toggle('hide', false)

        const order = store.get(constants.order)
        eventEmitter.payment.emit(constants.setPayment, this.type)

        order().op(ccConstants.ORDER_OP_INITIATE)
        order().handlePlaceOrder()
    }

    onChange = (options) => (state, component) => {
        const paymentDetails = store.get(constants.paymentDetails)
        const isValid = component.isValid && typeof state.data === 'object'
        eventEmitter.store.emit(constants.isValid, isValid)

        const payload = { ...paymentDetails, [this.type]: { ...state.data, ...options } }

        eventEmitter.store.emit(constants.paymentDetails, payload)
    }
}

export default Checkout
