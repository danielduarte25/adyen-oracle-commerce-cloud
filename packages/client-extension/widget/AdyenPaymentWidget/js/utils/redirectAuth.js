import storageApi from 'storageApi'
import * as constants from '../constants'
import { store } from '../components'
import { createFromAction } from './index'

export default ({ order, customPaymentProperties }, cb) => {
    const checkoutCard = store.get(constants.checkout.card)
    const isAuthRedirect = 'paymentData' in customPaymentProperties.additionalData
    const redirect = () => {
        const action = {
            action: customPaymentProperties.additionalData,
            selector: '#adyen-card-payment',
            checkoutComponent: checkoutCard,
        }

        const instance = storageApi.getInstance()
        instance.setItem(constants.storage.paymentData, customPaymentProperties.additionalData.paymentData)
        instance.setItem(constants.storage.order, JSON.stringify(order))

        createFromAction(action)
    }

    isAuthRedirect ? redirect() : cb(order)
}
