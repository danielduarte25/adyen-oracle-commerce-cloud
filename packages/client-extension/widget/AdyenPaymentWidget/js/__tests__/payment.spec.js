jest.mock('../components/static/bundle')

import Widget from '../../../../__mocks__/widget'
import viewModel from '../adyen-checkout'
import { Payment, store } from '../components'
import { eventEmitter } from '../utils'
import * as constants from '../constants'

const mockedPaymentDetails = { scheme: { paymentMethod: { type: 'scheme' } } }
const getPaymentResponse = (type, id) => [
    {
        customProperties: {
            numberOfInstallments: 3,
            paymentDetails: mockedPaymentDetails[type],
            storedPayment: undefined,
        },
        id,
        type: 'generic',
    },
]
describe('Payment', () => {
    let widget
    beforeEach(() => {
        widget = new Widget()
    })

    it('should create a payment', function () {
        viewModel.onLoad(widget)

        const payment = new Payment()
        const id = store.get(constants.id)
        const paymentMethodType = constants.paymentMethodTypes.scheme
        const paymentResponse = getPaymentResponse(paymentMethodType, id())

        eventEmitter.store.emit(constants.selectedInstallment, { numberOfInstallments: 3 })
        eventEmitter.store.emit(constants.genericPayment, paymentResponse[0])
        eventEmitter.store.emit(constants.paymentDetails, mockedPaymentDetails)
        payment.setPayment(constants.paymentMethodTypes.scheme)

        const order = store.get(constants.order)
        expect(order().payments()).toMatchSnapshot();
    })

    it('should have different payment method type on combo card', function () {
        viewModel.onLoad(widget)

        const payment = new Payment()
        const id = store.get(constants.id)
        const paymentMethodType = constants.paymentMethodTypes.scheme
        const paymentResponse = getPaymentResponse(paymentMethodType, id())
        const bin = constants.bins.electron

        eventEmitter.store.emit(constants.genericPayment, paymentResponse[0])
        eventEmitter.store.emit(constants.selectedComboCard, constants.comboCards.debit)
        eventEmitter.store.emit(constants.selectedBrand, bin)
        eventEmitter.store.emit(constants.paymentDetails, mockedPaymentDetails)
        payment.setPayment(constants.paymentMethodTypes.scheme)

        const order = store.get(constants.order)
        expect(order().payments()).toMatchSnapshot();
    })
})
