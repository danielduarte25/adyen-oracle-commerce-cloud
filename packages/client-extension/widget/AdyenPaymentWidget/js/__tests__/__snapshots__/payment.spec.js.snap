// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payment should create a payment 1`] = `
Array [
  Object {
    "customProperties": Object {
      "accountInfo": Object {
        "accountCreationDate": "01-01-1970",
      },
      "numberOfInstallments": 3,
      "paymentDetails": Object {
        "paymentMethod": Object {
          "type": "scheme",
        },
      },
      "riskData": Object {
        "basket": Object {
          "item1": Object {
            "brand": "mocked_brand",
            "category": "mocked_category",
            "itemID": "mocked_item_id",
            "productTitle": "mocked_product_title",
            "quantity": 1,
            "sku": "mocked_sku",
          },
        },
      },
      "storedPayment": undefined,
    },
    "id": "random_mocked_id",
    "type": "generic",
  },
]
`;

exports[`Payment should have different payment method type on combo card 1`] = `
Array [
  Object {
    "customProperties": Object {
      "accountInfo": Object {
        "accountCreationDate": "01-01-1970",
      },
      "additionalData": Object {
        "overwriteBrand": true,
      },
      "paymentDetails": Object {
        "paymentMethod": Object {
          "type": "electron",
        },
      },
      "riskData": Object {
        "basket": Object {
          "item1": Object {
            "brand": "mocked_brand",
            "category": "mocked_category",
            "itemID": "mocked_item_id",
            "productTitle": "mocked_product_title",
            "quantity": 1,
            "sku": "mocked_sku",
          },
        },
      },
      "selectedBrand": "electron",
      "storedPayment": undefined,
    },
    "id": "random_mocked_id",
    "type": "generic",
  },
]
`;
